### DISCUSSION TITLE
Submit action has been blocked by CORS policy

### DISCUSSION THREAD

**Discussion by StasWideweb** [solspace/craft-freeform#1438] [Questions & Answers] [answered]
The project is divided into two applications, one is responsible for the backend, the second for the frontend. Both applications are managed through one cloudways server. 
frontend build via gatsby and GraphQL.
GraphQL for the form
```
query GetAQuoteForm {
        craft {
          freeform {
            form(handle: "getAQuote") {
              id
              hash
              name
              handle
              csrfToken {
                name
                value
              }
              pages {
                label
                rows {
                  fields {
                    id
                    type
                    label
                    handle
                    required
                    ... on CraftCMS_FreeformField_Html {
                      value
                    }
                    ... on CraftCMS_FreeformField_Dropdown {
                      id
                      options {
                        label
                        value
                      }
                    }
                    ... on CraftCMS_FreeformField_Radios {
                      id
                      options {
                        label
                        value
                      }
                    }
                    ... on CraftCMS_FreeformField_Checkbox {
                      id
                      label
                      value
                    }
                  }
                }
              }
            }
          }
        }
      }
```
then inside js file we get info about form hash and csrf
```
useEffect(() => {
    const fetchFormData = async () => {
      try {
        const response = await fetch(
          `https://phpstack-example.cloudwaysapps.com/dynamic?form=${formData.handle}`
        )
        const result = await response.json()
        setCsrfToken(result.csrf.value)
        setFormHash(result.hash)
      } catch (error) {
        console.error(error)
      }
    }

    fetchFormData()
  }, [formData?.handle])
```
prepearing and send submit in the same js
```
const handleSubmit = async e => {
    e.preventDefault()

    if (validateForm()) {
      const formPayload = new URLSearchParams({
        ...inputValues,
        id: formData.id,
        [formData.csrfToken.name]: csrfToken,
        formHash: formHash,
      }).toString()

      console.log("csrfToken:" + csrfToken, "formHash" + formHash, formPayload)

      try {
        const response = await fetch(
          `https://phpstack-example.cloudwaysapps.com/actions/freeform/submit`,
          {
            method: "POST",
            headers: {
              "Content-Type": "application/x-www-form-urlencoded",
            },
            body: formPayload,
          }
        )

        if (response.ok) {
          const result = await response.json()
          console.log("Form submitted successfully:", result)
        } else {
          console.error("Form submission failed:", response.statusText)
        }
      } catch (error) {
        console.error("Error submitting form:", error)
      }
    }
  }
```
and get error when try to send submission
<img width="1672" alt="Screenshot at Jul 30 00-46-47" src="https://github.com/user-attachments/assets/f341a681-eb3d-426b-b2f4-f1c9accdc367">
also I've try to use EVENT_CONFIGURE_CORS inside default site Module
```
parent::init();

        Event::on(
            SubmitController::class,
            SubmitController::EVENT_CONFIGURE_CORS,
            function (ConfigureCORSEvent $event) {
                $headers = $event->getHeaders();

                $headers['Access-Control-Allow-Origin'] = '*';
                $headers['Access-Control-Allow-Methods'] = 'GET, POST, OPTIONS, PUT, DELETE';
                $headers['Access-Control-Allow-Headers'] = ['Content-Type, Authorization'];
                $headers['Access-Control-Allow-Credentials'] = 'true';

                $event->setHeaders($headers);
            }
        );
```
CORS headers enabled inside server app settings
https://prnt.sc/ifhkEyzT147z
P. S. 
I think after many hours spent trying to figure out why the submission failed, i ended up doing too many different things that are now possibly interfering with each other.

**Comment by seandelaney** [comment]
Hello @StasWideweb 

Have you taken a look at our headless demo repos? 

We have a number of working examples using Vue, React and NextJs for Freeform v4 and Freeform v5.

https://github.com/solspace/craft-freeform-demo-vuejs-graphql
https://github.com/solspace/craft-freeform-demo-vuejs-ajax

https://github.com/solspace/craft-freeform-demo-nextjs-graphql
https://github.com/solspace/craft-freeform-demo-nextjs-ajax

https://github.com/solspace/craft-freeform-demo-reactjs-graphql
https://github.com/solspace/craft-freeform-demo-reactjs-ajax

https://docs.solspace.com/craft/freeform/v4/headless
https://docs.solspace.com/craft/freeform/v5/headless

I don't think you should be playing around with CORS or anything else like this though, unless your frontend and backend are on different domains/subdomains.

**Reply by StasWideweb** [reply]
They are on the same server, but in different domains, which leads to error. I managed to fix the error directly CORS, when I removed all Cloudways settings and all the supporting code, I added the header settings directly to index.php, and it helped.
Now the error has changed to 400. Could this be because the CSRF token I get from the dynamic template is invalid by the time the form is submitted?

**Reply by seandelaney** [reply, marked as answer]
It's hard to say without seeing your code or the 400 error message - does it say unable to verify submission data?

In our demos, you'll see we grab the CSRF, hash, honeypot and payload values on page load (Example https://github.com/solspace/craft-freeform-demo-vuejs-ajax/blob/main/src/QuoteForm.vue#L43), and set these in the POST request when form is submitted. 

We also set these headers too:

`'X-CSRF-TOKEN': 'XYZ'`,
`'X-Requested-With': 'XMLHttpRequest'`
`'HTTP_X_REQUESTED_WITH': 'XMLHttpRequest'`

Double check everything. If still not working. Strip things back to a simple setup/form and get Craft accepting the POST request first.
